### ISSUE TITLE
tus Concatenation Extension Proposal

### ISSUE THREAD

**Issue by ccampbell** [tus/tus-resumable-upload-protocol#63]
After reviewing the spec for the concatenation extension I think that the
current proposal can be simplified a bit.

Right now each client is required to make individual file creation requests with
individual content lengths for each segment, keep track of all of them and the
order they should be in, and then send another request with the list of the
individual segment urls.

This puts a bit of a burden on the client implementation that can be avoided if
the server does more of the work.

### Proposal

Have the client make a file creation request with the final `Upload-Length` and
a new header `Upload-Segments` which would be an integer greater than 0, but
less than the total Upload-Length.

If the number of requested segments seems too high the server can return an error
response.

The server would reply with a response containing a header with the URIs of all
of the segments the client will need to send data to in the order they should
appear possibly containing the total length of each segment as well.

It could also return the normal `Location` response with the location for the
finished upload which you could use HEAD against to get the overall status based
on the status of the individual segments.

### Why?

- Having to manage parallel requests to create individual segments is cumbersome
and might lead to strange bugs.  In javascript particularly doing parallel
asynchronous stuff can be annoying.
- It cuts down on the number of http requests.  In the current proposal if you
have 10 segments, it will require 11 HTTP requests just for file creation (not
including PATCH requests and HEAD requests).  In this new proposal, it only
requires 1 HTTP request
- Currently the server has to wait on an `Upload-Concat: final` and deal with
the possibility of it coming in before the uploads are finished using the
`concatenation-unfinished` extension.  That will no longer be necessary

### Example

This is an example of what the creation request might look like.  Not sure if
this is the best way to represent the expected segments and their sizes, but it
will require the least amount of parsing on the client cause you can just split
at comma.

The expected ranges/sizes technically do not need to be returned by this request
if there is an agreed upon way to calculate the segment sizes.

##### Possible calculation example

```javascript
sizePerSegment = Math.floor(uploadLength / numSegments)

segmentSize = sizePerSegment
if isLastSegment
    segmentSize = sizePerSegment + uploadLength % sizePerSegment
```

- Initialize the upload

    **Request**

    ```
    POST /files HTTP/1.1
    Upload-Length: 1005
    Upload-Segments: 10
    ```

    **Response**

    ```
    HTTP/1.1 201 Created
    Location: https://tus.example.org/files/5c8da61d68591b1300a46e3e2766daef
    Upload-Segments: /files/5c8,/files/a3e,/files/f87,/files/e4e,/files/cc5,/files/7d2,/files/88a,/files/13b,/files/b3b,/files/6a4
    Upload-Expected-Ranges: 0-99,100-199,200-299,300-399,400-499,500-599,600-699,700-799,800-899,900-1004
    ```

- Patch the expected segment to each of the segment URIs
- Use `HEAD https://tus.example.org/files/5c8da61d68591b1300a46e3e2766daef` to check overall state or `HEAD /files/cc5` to check individual segment state.

### Note

- This assumes that every segment will be an equal length, but I can't
think of a case where you would want to use different lengths for different
segments.
- Originally I had this for the response header, but I realized the ranges would probably reduce the amount of work required on the client

    ```
    Upload-Expected-Sizes: 100,100,100,100,100,100,100,100,100,105
    ```

**Comment by vayam** [comment]
@Acconut I see this possible simpler extension for uploading a file of known length in parallel chunks.

**Comment by Acconut** [comment]
First of all I want to thank you for your effort. Secondly, you may have a look at #55 where we first discussed how to approach parallel uploads.

> Having to manage parallel requests to create individual segments is cumbersome and might lead to strange bugs. In javascript particularly doing parallel asynchronous stuff can be annoying.

In my experience there are enough tools to help with parallel asynchronous tasks, whereas my preferred way is [async](https://www.npmjs.com/package/async).

> It cuts down on the number of http requests. In the current proposal if you have 10 segments, it will require 11 HTTP requests just for file creation (not including PATCH requests and HEAD requests). In this new proposal, it only requires 1 HTTP request

Good point.

> Currently the server has to wait on an `Upload-Concat: final` and deal with the possibility of it coming in before the uploads are finished using the `concatenation-unfinished` extension. That will no longer be necessary

The `concatenation-unfinished` extension introduces a problem for the server. Basically, once an upload is completed is has to check whether it is a partial one and if the corresponding final upload should be considered finished, too. In the case of [tusd's](https://github.com/tus/tusd) design this is not very practically. This is my biggest concern with this approach.

In addition, your solution removes the ability of uploading non-contiguous chunks since the segments have the same length. To be honest, I am not sure if this is a problem but we should consider this.

In the end, it's a question about who should do more work, the server or the client.

**Comment by Acconut** [comment]
Another point is that your solution does not allow to concatenate upload with an unknown length. Again, I cannot think of a real world case but we have to consdier this.

**Comment by vayam** [comment]
@Acconut @ccampbell  It solves the problem of uploading a large file in parallel.  There are few cases (transcode) where I would want to compose from non-contiguous variable length chunks. But this idea is much simpler for 90% of the use cases. May be worth exploring this as a separate extension.

**Comment by Acconut** [comment]
I am going to close this proposal since this Concatenation extension has been included in the 1.0 release. If you still want to discussion on this issue, feel free to comment and we will reopen it.
